### DISCUSSION TITLE
Revamping the Spring Boot Migrator (SBM) Project

### DISCUSSION THREAD

**Discussion by fabapp2** [spring-projects-experimental/spring-boot-migrator#859] [Announcements]
Hello,

We'd like to share the latest news on the Spring Boot Migrator (SBM) project. We've fallen behind in integrating recent [OpenRewrite](https://github.com/openrewrite/rewrite) versions and we apologize if this has affected you. Our team has gone through some changes and the workload has increased, leading to these delays.

During this time, the OpenRewrite engineering team launched a major update, now working on an 8.x baseline. Meanwhile, the current SBM version, 0.14.0, is still operating on the outdated OpenRewrite version 7.35.0.
Integrating the new release and addressing previously delayed features like [scoped beans](https://github.com/spring-projects-experimental/spring-boot-migrator/pull/792) are demanding tasks we're tackling.

## OpenRewrite Parser

We've seen other teams create their custom parsers for OpenRewrite recipes. Therefore, we decided to collaborate and create a shared parser component compatible with the latest Boot and OpenRewrite versions. You can find more details about this in this [Issue](https://github.com/spring-projects-experimental/spring-boot-migrator/issues/844) and the shared component is located under [`/sbm-support-rewrite`](https://github.com/spring-projects-experimental/spring-boot-migrator/tree/main/sbm-support-rewrite).

This parser simplifies:

- Project scanning.
- Parsing resources into OpenRewrite AST (for both Gradle and Maven).
- Discovering existing recipes.

*Parse, discover, and run OpenRewrite recipes:*
````java
Path baseDir = ...  
// sbm parser
ProjectParsingResult result = projectParser.parse(baseDir);
// provides access to OR AST
List<SourceFile> ast = result.getSourceFiles();
// and ExecutionContext
ExecutionContext ctx = result.getExecutionContext();
// It discovers OR recipes
Recipe recipe = discovery.findRecipe("...");
// And the AST and ExecutionContext 
// can then be used to execute the OR recipe
LargeSourceSet lss = InMemoryLargeSourceSet(ast);
RecipeRun result = recipe.apply(lss, ctx);
````

## Introducing the Parser
The upcoming release 0.15.0 will be the last on the 7.35.0 OpenRewrite version and the last version before we integrate the new parser. 
This update will affect all dependent modules, so we'll decouple all modules from the root pom and upgrade each one separately. 

### Replace `sbm-core` Parser

The first step is updating `sbm-core` to use the new parser, replacing the `RewriteMavenProjectParser`. The progress is tracked in [this issue](https://github.com/spring-projects-experimental/spring-boot-migrator/issues/858).

The new `sbm-core` will apply OpenRewrite recipes, handle changes efficiently, sync seamlessly with the filesystem, and optionally support GIT.

*Run OpenRewrite recipes through ProjectContext:*
````java
@Component
public class RecipeExecutor {
	@Autowired
	ProjectContextInitializer initializer;
	@Autowired
	RewriteRecipeDiscovery discovery;
	@Autowired
	ProjectContextSerializer serializer;
		
	public void runOpenRewriteRecipe(Path baseDir, String name) {
		Recipe recipe = discovery.discoverRecipe(name);
		ProjectContext pc = initializer.initProjectContext(baseDir);
		pc.getProjectJavaSources().apply(recipe);
		serializer.writeChanges(pc); 	
	} 
}
````

### Enhance `ProjectContext`

After updating `sbm-core`, we aim to improve `ProjectContext` by introducing a method to run existing or new OpenRewrite recipes.

Recipes can then be executed like this:

````java
ProjectContext pc = scanCommand.execute(baseDir); pc.apply(new OneOpenRewriteRecipe());
pc.apply(new AnotherOpenRewriteRecipe());
serializer.writeChanges(pc);
pc.apply(new TheLastRecipe());
serializer.writeChanges(pc);
````

This makes creating custom tools for OpenRewrite recipes easier.

### Upgrade Select SBM Recipes

Finally, we plan to make selected SBM recipes compatible with OpenRewrite 8.x, which will involve migrating modules hosting these recipes, like `sbm-recipes-jee-to-boot` or `sbm-recipes-boot-upgrade`.

## SBM Shell and Boot Upgrade

This transition lets us improve several components. We plan to simplify the Spring Boot Upgrade Report and find the best execution strategy.

We'll discontinue the SBM shell application under `applications/spring-shell` and incorporate important components into [SpringCLI](https://github.com/spring-projects-experimental/spring-cli). This will provide a convenient platform for executing SBM, OpenRewrite, and other recipes.

Thanks for your support during these updates. We look forward to your continued feedback. Feel free to reach out if you have any questions or comments!

**Comment by anbusampath** [comment]
is [[spring-rewrite-commons](https://github.com/spring-projects/spring-rewrite-commons)] to revamping?

**Reply by addozhang** [reply]
How about the progress of revamping now?
